sound/soc/sof/pcm.c

```diff
@@ -689,6 +689,83 @@ static snd_pcm_sframes_t sof_pcm_delay(struct snd_soc_component *component,
 	return 0;
 }
 
+static int sof_pcm_trigger_suspended_paused_streams(struct snd_sof_dev *sdev,
+						    int cmd)
+{
+	struct snd_pcm_substream *substream;
+	struct snd_pcm_runtime *runtime;
+	struct snd_sof_pcm *spcm;
+	int dir, ret;
+
+	list_for_each_entry(spcm, &sdev->pcm_list, list) {
+		for_each_pcm_streams(dir) {
+			substream = spcm->stream[dir].substream;
+			if (!substream || !substream->runtime)
+				continue;
+
+			/*
+			 * The stream supports RESUME, it is expected that it
+			 * is handling the corner case of suspending while
+			 * a stream is paused
+			 */
+			runtime = substream->runtime;
+			if (runtime->info & SNDRV_PCM_INFO_RESUME)
+				continue;
+
+			/* Only send the trigger to a paused and suspended stream */
+			if (runtime->state != SNDRV_PCM_STATE_SUSPENDED ||
+			    runtime->suspended_state != SNDRV_PCM_STATE_PAUSED)
+				continue;
+
+			ret = substream->ops->trigger(substream, cmd);
+			if (ret) {
+				dev_err(sdev->dev,
+					"%s: trigger %d failed for stream %d, dir: %d\n",
+					__func__, cmd, spcm->pcm.pcm_id, dir);
+				return ret;
+			}
+		}
+	}
+
+	return 0;
+}
+
+int sof_pcm_stop_paused_on_suspend(struct snd_sof_dev *sdev)
+{
+	int ret;
+
+	/*
+	 * Handle the corner case of system suspend while at least one stream is
+	 * paused.
+	 * Paused streams will not receive the SUSPEND triggers, they are
+	 * 'silently' moved to SUSPENDED state.
+	 *
+	 * The workaround for the corner case is applicable for streams not
+	 * supporting RESUME.
+	 *
+	 * First we need to move (trigger) the paused streams to RUNNING state,
+	 * then we need to stop them
```

Comment:
why is this necessary? it's perfectly legal to go from PAUSED to STOPPED. it's not required to go back to RUNNING

This is the case for soc-pcm.c:
````c
		case SNDRV_PCM_TRIGGER_STOP:
			if ((be->dpcm[stream].state != SND_SOC_DPCM_STATE_START) &&
			    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_PAUSED))
				goto next;
`````

and likewise IPC4 transition are fine with a PAUSE->STOP.

IMHO the transition from PAUSED to RUNNING creates a racy behavior where the state can change and some samples might be generated before stopping again.

Reply:
> why is this necessary? it's perfectly legal to go from PAUSED to STOPPED. it's not required to go back to RUNNING
> 
> This is the case for soc-pcm.c:
> 
> ```c
> 		case SNDRV_PCM_TRIGGER_STOP:
> 			if ((be->dpcm[stream].state != SND_SOC_DPCM_STATE_START) &&
> 			    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_PAUSED))
> 				goto next;
> ```

@plbossart, it might be so in ASoC DPCM, but if a stream is PAUSED then the core would not send a STOP trigger:
```c
static int snd_pcm_do_stop(struct snd_pcm_substream *substream,
			   snd_pcm_state_t state)
{
	if (substream->runtime->trigger_master == substream &&
	    snd_pcm_running(substream)) {
		substream->ops->trigger(substream, SNDRV_PCM_TRIGGER_STOP);
		substream->runtime->stop_operating = true;
	}
	return 0; /* unconditionally stop all substreams */
}
```
 
> and likewise IPC4 transitions are fine with a PAUSE->STOP.

Furthermore, we in sof do not handle well the PAUSE->STOP if we handle it all.
Yes, I have tested this as the first thing.

> IMHO the transition from PAUSED to RUNNING creates a racy behavior where the state can change and some samples might be generated before stopping again.

Yes, that might happen. We are talking about a rare corner case. The options is to end up with a broken stack or a possible glitch.